src/cli/run_command.zig

```diff
@@ -227,7 +234,387 @@ pub const RunCommand = struct {
+
+            const shell_bin = findShell(env.map.get("PATH") orelse "", cwd) orelse return error.MissingShell;
+
+            var copy_script = try std.ArrayList(u8).initCapacity(manager.allocator, original_script.script.len + 1);
+            try replacePackageManagerRun(&copy_script, original_script.script);
+            try copy_script.append(0);
+
+            var combined_script: [:0]u8 = copy_script.items[0 .. copy_script.items.len - 1 :0];
+
+            var argv = try manager.allocator.allocSentinel(?[*:0]const u8, 3, null);
+            defer manager.allocator.free(argv);
+            argv[0] = shell_bin;
+            argv[1] = "-c";
+            argv[2] = combined_script;
+
+            var flags: i32 = bun.C.POSIX_SPAWN_SETSIGDEF | bun.C.POSIX_SPAWN_SETSIGMASK;
+            if (comptime Environment.isMac) {
+                flags |= bun.C.POSIX_SPAWN_CLOEXEC_DEFAULT;
+            }
+
+            var attr = try PosixSpawn.Attr.init();
+            defer attr.deinit();
+            try attr.set(@intCast(flags));
+            try attr.resetSignals();
+
+            var actions = try PosixSpawn.Actions.init();
+            defer actions.deinit();
+            try actions.openZ(bun.STDIN_FD, "/dev/null", std.os.O.RDONLY, 0o664);
+
+            // Have both stdout and stderr write to the same buffer
+            const fdsOut = try std.os.pipe2(0);
+            try actions.dup2(fdsOut[1], bun.STDOUT_FD);
+
+            const fdsErr = try std.os.pipe2(0);
+            try actions.dup2(fdsErr[1], bun.STDERR_FD);
+
+            try actions.chdir(cwd);
+
+            var arena = bun.ArenaAllocator.init(manager.allocator);
+            defer arena.deinit();
+
+            const pid = brk: {
+                defer {
+                    _ = bun.sys.close(fdsOut[1]);
+                    _ = bun.sys.close(fdsErr[1]);
+                }
+                switch (PosixSpawn.spawnZ(
+                    argv[0].?,
+                    actions,
+                    attr,
+                    argv,
+                    try env.map.createNullDelimitedEnvMap(arena.allocator()),
+                )) {
+                    .err => |err| {
+                        Output.prettyErrorln("<r><red>error<r>: Failed to spawn script <b>{s}<r> due to error <b>{d} {s}<r>", .{ name, err.errno, @tagName(err.getErrno()) });
+                        Output.flush();
+                        return;
+                    },
+                    .result => |pid| break :brk pid,
+                }
+            };
+
+            const pid_fd: std.os.fd_t = brk: {
+                if (!Environment.isLinux) {
+                    break :brk pid;
+                }
+
+                const kernel = @import("../analytics.zig").GenerateHeader.GeneratePlatform.kernelVersion();
+
+                // pidfd_nonblock only supported in 5.10+
+                const pidfd_flags: u32 = if (kernel.orderWithoutTag(.{ .major = 5, .minor = 10, .patch = 0 }).compare(.gte))
```

Comment:
Can you extract this into a helper method that is shared between subprocess.zig and this code?